### PR TITLE
Add support for Vulkan timeline semaphores.

### PR DESCRIPTION
**Pull request by yalcinmelihyasin**
This PR is a cherry-pick from the GAPID PR #3791 and a few changes to fix
merge conflicts.

Manually tested with Vulkan test applications.

b/174312320

### CODE REVIEW

**Comment by yalcinmelihyasin** [comment]
Please review but *do not merge*

The tests are not complete yet, I'll merge when I am sure with all the tests.

File: gapis/api/vulkan/externs.go [hunk 1]
```diff
@@ -108,19 +108,37 @@ func (e externs) onCommandAdded(buffer VkCommandBuffer) {
...
+	o.SubCmdIdx = append(api.SubCmdIdx{}, o.LastSubCmdIdx...)
+	delete(o.deferredSubmissions, ref)
+}
+
+func (e externs) enterSubcontext() {
+	o := GetState(e.s)
+	if len(o.SubCmdIdx) == 0 {
+		o.SubCmdIdx = api.SubCmdIdx{uint64(e.cmdID)}
+	}
+	o.SubCmdIdx = append(o.SubCmdIdx, 0)
```
Comment (hevrard) [inline gapis/api/vulkan/externs.go:133]: This change breaks the semantics of o.SubCmdIdx: so far, it did not contain the top-level parent command ID, now it does. This is why we have the corresponding change inside MutateSubcommands(), in the sub command index argument passed in the call backs. Now I hope nothing else was relying on o.SubCmdIdx being only the tail of the command indexes.

**Review by hevrard — COMMENTED** [review]
Thanks for this! The logic LGTM, just a few cleanups on comments / formatting.

Nit on referring another PR in a commit message: you can use the full github PR URL (e.g. `https://github.com/google/gapid/pull/3791`) instead of `GAPID PR #3791`, the github UI automatically presents the URL in a shorter form: https://github.com/google/gapid/pull/3791. This makes it easier to visit the PR when looking at the commit, both in github and in other git interfaces.

File: gapis/api/sync/data.go [hunk 2]
```diff
@@ -95,6 +95,11 @@ type Data struct {
 	// created to organize psubmits, command buffers, etc.
 	SubcommandNames   *api.SubCmdIdxTrie
 	SubmissionIndices map[api.CmdSubmissionKey][]api.SubCmdIdx
+	// UnblockingCommands is a map of what command fully unblocks
+	// some other command. Or to put it another way, the command that answers
+	// the question: If we want to play up to command N, what command do we
+	// have to replay until, in order to prevent a deadlock.
+	UnblockingCommands map[api.CmdID]api.CmdID
```
[the remark this answers is not in the document]
Comment (AWoloszyn) [reply gapis/api/sync/data.go:102]: SyncData *NEVER* includes initial commands. In fact initial commands are simply a byproduct of the replay. They do not "exist" anywhere outside replay.  A CmdID is very well defined outside of replay, so we should probably NOT add additional confusion here by talkign about them.

[the remark this answers is not in the document]
Comment (hevrard) [reply gapis/api/sync/data.go:102]: OK, thanks for the clarification.

**Comment by yalcinmelihyasin** [comment]
Closing for now until sufficient device support.

**Pull request closed without merging**